### PR TITLE
[WIP] Jrs/replay buffer

### PR DESCRIPTION
**Pull request by jrs65**
@tristpinsm this is the WIP branch that does a few refactors. Note, that I haven't tested my changes yet. It seems that the only unit test that uses `visRawReader` is for `visTranspose` so I guess they'll need to be changed together before they pass.

### CODE REVIEW

File: lib/stages/visRawReader.cpp [hunk 1]
```diff
@@ -303,64 +322,52 @@ void visRawReader::main_thread() {
...
             frame.fpga_seq_length = 0;
+            frame.fpga_seq_total = 0;
+
             DEBUG("visRawReader: Reading empty frame: {:d}", frame_id);
         }
 
         // Set the dataset ID to the updated value
-        ((visMetadata*)(out_buf->metadata[frame_id]->metadata))->dataset_id = out_dset_id;
+        dset_id_t& ds_id = ((visMetadata*)(out_buf->metadata[frame_id]->metadata))->dataset_id;
+        ds_id = get_dataset_state(ds_id);
```
Comment (tristpinsm) [inline lib/stages/visRawReader.cpp:353]: I guess this means it will looks through the keys of the `ds_in_file` map for every frame. This is probably a small overhead, but I don't have a clear idea of how many different dataset IDs there can be in one file.

**Review by tristpinsm — COMMENTED** [review]
This looks good to me so far. It's also made the handling of dataset IDs quite a bit clearer... After so many iterations it had gotten pretty messy.

**Comment by tristpinsm** [comment]
Should I patch `visTranspose` on top of this branch and make a PR onto here then? I could also try and extend the unit test a bit to include empty frames and check the dataset IDs.

File: lib/stages/visRawReader.cpp (hunk 1 above)
Comment (jrs65) [reply lib/stages/visRawReader.cpp:353]: Yeah. I figured it was worth the very small amount of overhead to clarify the handling. Lookups in a `std::map` are `O(log N)` so pretty cheap, and at most we might have 10s of datasets per file, and 100s if we go to very fast cadence updates.

**Comment by jrs65** [comment]
> Should I patch `visTranspose` on top of this branch and make a PR onto here then? I could also try and extend the unit test a bit to include empty frames and check the dataset IDs.

Maybe a PR on top of this make sense. Extending the unit tests while you do so would be great.

**Comment by jrs65** [comment]
@andrerenard this is still a thing. I'll try and review it's final status more in the near future.

**Comment by andrerenard** [comment]
Closing due to age, and similar work in current code.
